swift/ql/src/queries/Security/CWE-916/InsufficientHashIterations.qhelp

```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+  <overview>
+    <p>Using hash functions with less than 1,000 iterations is not secure. That scheme is vulnerable to password cracking attacks due to having an insufficient level of computational effort.</p>
+  </overview>
+
+  <recommendation>
+    <p>Use sufficient number of iterations (i.e., greater than or equal 1000) for generating password-based keys.</p>
```

Comment:
```suggestion
    <p>Use sufficient number of iterations (that is, greater than or equal 1000) for generating password-based keys.</p>
```
I believe "that is," is preferred over "i.e." (for readers with less experience in English).

Out of interest, I understand this comes from previous research but I'm interested to hear why the threshold we chose is 1,000, rather than say 100 or 10,000?  Is this recommendation reasonably future proof or is it something we might need to increase in future?

Reply:
Good question, the initial standard is 1000, but that number should increase with more CPU power (https://en.wikipedia.org/wiki/PBKDF2). The latest recommendation from OWASP is actually at least 120,000 iterations (https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html). Given that this number should increase over time, how do we best handle this?

Reply:
If we want to increase it over time, I guess we should create some kind of issue to track this and preferably attach it to something (like an iteration or quarter) far in the future.

But it sounds like there's already a big gap between the current threshold of 1,000 and the OWASP recommendation of 120,000 iterations.  I guess we don't want to be too fussy - telling someone that 100,000 iterations isn't enough when the recommendation is 120,000 seems picky - but we do want to call out when code is risky e.g. because it follows advice from 20 years ago.  Is there any reason we shouldn't increase our threshold to somewhere in the 10,000 - 100,000 range?

Reply:
... in fact reading the OWASP page, 120,000 appears to be the lowest of the recommendations they make.  So perhaps we _should_ set the threshold to 120,000???

Reply:
I'd say 120k should be the lowest threshold that we should generate warnings for. I'll update the queries and the associated documentation, examples, etc.